lib/Sema/SemaBounds.cpp

```diff
@@ -520,22 +518,77 @@ namespace {
             assert("unexpected cast failure");
             return CreateBoundsNone();
           }
-          switch (UO->getOpcode()) {
-            case UnaryOperatorKind::UO_AddrOf: {
-              Expr *SubExpr = UO->getSubExpr();
-              if (SubExpr->getType()->isFunctionType())
-                return CreateBoundsNone();
+          UnaryOperatorKind Op = UO->getOpcode();
 
-              return LValueBounds(SubExpr);
-            }
-            default:
-              // TODO: fill in other cases
+          if (Op == UnaryOperatorKind::UO_AddrOf) {
+            Expr *SubExpr = UO->getSubExpr();
+            if (SubExpr->getType()->isFunctionType())
               return CreateBoundsNone();
+
+              return LValueBounds(SubExpr);
           }
+
+          if (UnaryOperator::isIncrementDecrementOp(Op))
+            return LValueTargetBounds(UO->getSubExpr());
+
+          if (Op == UnaryOperatorKind::UO_Plus ||
+              Op == UnaryOperatorKind::UO_Minus ||
+              Op == UnaryOperatorKind::UO_Not)
+            return RValueBounds(UO->getSubExpr());
+          return CreateBoundsNone();
         }
-        // TODO: fill in these cases
         case Expr::BinaryOperatorClass:
-        case Expr::CompoundAssignOperatorClass:
+        case Expr::CompoundAssignOperatorClass: {
+          BinaryOperator *BO = dyn_cast<BinaryOperator>(E);
+          if (!BO) {
+            assert("unexpected cast failure");
+            return CreateBoundsNone();
+          }
+          Expr *LHS = BO->getLHS();
+          Expr *RHS = BO->getRHS();
+          BinaryOperatorKind Op = BO->getOpcode();
+
+          if (Op == BinaryOperatorKind::BO_Assign)
+            return RValueBounds(RHS);
+
+          bool IsCompoundAssignment = false;
+          if (BinaryOperator::isCompoundAssignmentOp(Op)) {
+            Op = BinaryOperator::getOpForCompoundAssignment(Op);
+            IsCompoundAssignment = true;
+          }
+
+          // Pointer arithmetic.
+          if (LHS->getType()->isPointerType() &&
+              RHS->getType()->isIntegerType() &&
+              BinaryOperator::isAdditiveOp(Op)) {
+            return IsCompoundAssignment ?
+              LValueTargetBounds(LHS) : RValueBounds(LHS);
+          }
+          if (RHS->getType()->isPointerType() &&
+              LHS->getType()->isIntegerType() &&
```

Comment:
If it confused you, it'll confuse other people.  I'll change the code so that the operands are tested in the same order as the prior test.